File: src/main/java/pl/training/workshop/m7/s08_designbycontract/start/SeatPool.java
package pl.training.workshop.m7.s08_designbycontract.start;

/**
 * Start: pula miejsc seansu bez żadnych kontraktów. reserve(-2) po cichu "dodaje" miejsca,
 * a release() pozwala przekroczyć pojemność sali. Stan może stać się niemożliwy.
 */
public final class SeatPool {
    private final int capacity;
    private int remaining;

    public SeatPool(int capacity) {
        this.capacity = capacity;
        this.remaining = capacity;
    }

    /** Zwraca false, gdy wolnych miejsc jest za mało - to poprawna, udokumentowana odpowiedź. */
    public boolean reserve(int seats) {
        if (seats > remaining) {
            return false;
        }
        remaining = remaining - seats;
        return true;
    }

    public void release(int seats) {
        remaining = remaining + seats;
    }

    public int remaining() {
        return remaining;
    }

    public int capacity() {
        return capacity;
    }
}

File: csharp/src/Training.Workshop/M7/S08DesignByContract/Start/SeatPool.cs
namespace Training.Workshop.M7.S08DesignByContract.Start;

/// <summary>
/// Start: pula miejsc seansu bez żadnych kontraktów. Reserve(-2) po cichu "dodaje" miejsca,
/// a Release() pozwala przekroczyć pojemność sali. Stan może stać się niemożliwy.
/// </summary>
public sealed class SeatPool
{
    private readonly int _capacity;
    private int _remaining;

    public SeatPool(int capacity)
    {
        _capacity = capacity;
        _remaining = capacity;
    }

    /// <summary>Zwraca false, gdy wolnych miejsc jest za mało - to poprawna, udokumentowana odpowiedź.</summary>
    public bool Reserve(int seats)
    {
        if (seats > _remaining)
        {
            return false;
        }
        _remaining = _remaining - seats;
        return true;
    }

    public void Release(int seats)
    {
        _remaining = _remaining + seats;
    }

    public int Remaining => _remaining;

    public int Capacity => _capacity;
}
